### PR TITLE
Update focal CN file prep to use exons again and cover the consensus SEG case

### LINKED ISSUE #473: Annotated CNVkit data no longer shows SMARCB1 deletions in ATRT
Context is rerunning the `molecular-subtyping-ATRT` module as part of #462.

#### What data file(s) does this issue pertain to?

`analyses/focal-cn-file-preparation/cnvkit_annotated_cn_autosomes.tsv.bz2` in the repo, not in the data release.

#### What release are you using?

The file was generated using `release-v13-20200116`

#### Put your question or report your issue here.

There are no longer _SMARCB1_ losses noted for the majority of ATRT samples since #452. You can see a notebook comparing the old annotated CNVkit file to the new CNVkit file here: https://jaclyn-taroni.github.io/openpbta-notebook-concept/segment-to-gene-mapping.nb.html

I assume that our finding of overlaps or filtering is more restrictive now. Here are a few thoughts as to why:

* Using the CDS BED file -- CDS should be more restrictive than exons
* Filtering based on cytoband (I don't think this is the major issue)
https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/78bba86523e150789b3929ff01bac129b689f3f7/analyses/focal-cn-file-preparation/01-prepare-cn-file.R#L92
* Using `findOverlaps`: https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/78bba86523e150789b3929ff01bac129b689f3f7/analyses/focal-cn-file-preparation/01-prepare-cn-file.R#L84
  instead of `mergeByOverlaps`: https://github.com/jaclyn-taroni/OpenPBTA-analysis/blob/fa214291713575be7fd20c92374b268870f4173f/analyses/focal-cn-file-preparation/01-prepare-cn-file.R#L82

It was previously suggested (https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/182#discussion_r340587968) to use _SMARCB1_ losses in ATRT to check if things were as expected.

As a side note, there are "duplicate" rows because of how the arm values were joined, e.g., there is information for every arm joined to a gene, not just the relevant arm. It makes sense to look into and try to address this at the same time.

### PR DESCRIPTION
**Pull request by jaclyn-taroni**
### Purpose/implementation Section

We used to use exons to annotate CNV data files the prior to #452, which made the switch to using coding regions. That looked like it was too restrictive and inconsistent with what is typically done (some references and quotes supplied by @cbethell below).

[Monlong et al. _NAR._ 2018.](https://doi.org/10.1093/nar/gky538):

> Exons of protein-coding genes and promoter regions (10 kb upstream of the transcription start site) were extracted from the Gencode annotation v19. We counted how many genes overlapped a CNV in the population when considering exons only, exons and promoter region, or gene body and promoter region

[Bailey et al. _Cytogenet Genome Res._ 2009.](https://dx.doi.org/10.1159/000184713):

> Our detection strategy was developed as a simple threshold-per-exon test with chaining of exons to predict the extent of the variation.

#### What was your approach?

Here we are: 
* Adding a notebook that prepares the consensus SEG file for the annotation step (`analyses/focal-cn-file-preparation/02-add-ploidy-consensus.Rmd`). There is a _lot_ of overlap with how we prepare the CNVkit file, so this can probably be refactored in a subsequent pull request. This basically means this code was reviewed before in #253 and #259.
* Essentially _reverting_ to a version of the `analyses/focal-cn-file-preparation/03-prepare-cn-file.R` file prior to #452, with some minor changes (making the option name for SEG files more general): https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/2af3f350a247d66f68fe9305224d937ee0e90a8c/analyses/focal-cn-file-preparation/01-prepare-cn-file.R  (This script has been renumbered as part of this pull request.)
* Adding a step to the shell script where we are annotating the consensus SEG file that's in `master` at the moment. We're also skipping the expression step currently.
* Renumbering scripts and notebooks.
* Reverting to writing gzipped files means that the downstream modules (and their documentation in the modules at a glance table) needed to be updated as well.

#### What GitHub issue does your pull request address?

Closes #473; related to #186 

#### Which areas should receive a particularly close look?

This looks a bit worse than it is, as most of the code has gone through review before (sorry!)

#### Reproducibility Checklist

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

#### Documentation Checklist

- [x] This analysis module has a `README` and it is up to date.
- [x] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

### CODE REVIEW

**Comment by jaclyn-taroni** [comment]
Rendered version of the notebook here: https://jaclyn-taroni.github.io/openpbta-notebook-concept/02-add-ploidy-consensus.nb.html

File: analyses/focal-cn-file-preparation/03-prepare-cn-file.R [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+# https://github.com/marislab/pptc-pdx-copy-number-and-SVs/blob/master/R/focal-CN-revision.R
+#
+# Chante Bethell for CCDL 2019 and Jo Lynne Rokita
+#
+# #### USAGE
+# This script is intended to be run via the command line from the top directory
+# of the repository as follows:
+#
+# Rscript 'analyses/oncoprint-landscape/01-prepare-cn-file.R'
+
```
Comment (cansavvy) [inline analyses/focal-cn-file-preparation/03-prepare-cn-file.R]: Unless this script would work without specifying anything,  I suspect this usage example is incomplete?

File: analyses/focal-cn-file-preparation/02-add-ploidy-consensus.Rmd [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+                    germline_sex_estimate), 
+             by = c("ID" = "Kids_First_Biospecimen_ID")) %>%
+  select(-tumor_ploidy, -germline_sex_estimate, everything())
+```
+
+### Handle autosomes first
+
+```{r}
+add_autosomes_df <- add_ploidy_df %>%
+  # x and y chromosomes must be handled differently - leave in for safety
```
Comment (cansavvy) [inline analyses/focal-cn-file-preparation/02-add-ploidy-consensus.Rmd]: leave in for safety?

File: analyses/focal-cn-file-preparation/03-prepare-cn-file.R [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+                                      txdb_exons,
+                                      filt_na_symbol = TRUE) {
+  # This function takes a standardized data.frame that contains genomic range
+  # information and finds the overlaps with a TxDb object. It will discard
+  # Ensembl gene IDs with no gene symbol by default.
+  #
+  # Args:
+  #   cnv_df: standardized data.frame that contains the segments used in the
+  #           CNV caller
+  #   txdb_exons: exons to be merged with the CNV data.frame; output o
```
Comment (cansavvy) [inline analyses/focal-cn-file-preparation/03-prepare-cn-file.R]:
```suggestion
  #   txdb_exons: exons to be merged with the CNV data.frame; output of
```

File: analyses/focal-cn-file-preparation/03-prepare-cn-file.R [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+  file.path(root_dir, "analyses", "focal-cn-file-preparation", "results")
+
+if (!dir.exists(results_dir)) {
+  dir.create(results_dir)
+}
+
+#### Format CNV file and overlap with hg38 genome annotations ------------------
+
+# we want to standardize the formats between the two methods here and drop
+# columns we won't need to
```
Comment (cansavvy) [inline analyses/focal-cn-file-preparation/03-prepare-cn-file.R]:
```suggestion
# columns we won't need.
```

File: analyses/focal-cn-file-preparation/README.md [hunk 5]
```diff
@@ -17,26 +15,31 @@ bash analyses/focal-cn-file-preparation/run-prepare-cn.sh
...
   This notebook adds the ploidy inferred via ControlFreeC to the CNVkit data and adds a status column that defines gain and loss broadly.
   (Note that [the logic around sex chromosomes in males when ploidy = 3 leaves something to be desired](https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/259#discussion_r345354403)).
-* `01-prepare-cn-file.R` - This script performs the ranges to annotation mapping using the GENCODE v27 GTF included via the data download step; it takes the ControlFreeC file and the CNVkit file prepared with `00-add-ploidy-cnvkit.Rmd` as input.
-  The mapping is limited to _coding sequences_.
-  Mapping to cytobands is performed with the [`UCSC hg38 cytoband file`](http://hgdownload.cse.ucsc.edu/goldenpath/hg38/database/cytoBand.txt.gz).
-  _Note: The decision to implement the `UCSC file` was made based on a comparison done between the cytoband calls in the `org.Hs.eg.db` package and the calls in the `UCSC file`. We found that they disagreed in ~11,800 calls out of ~800,000 and the `UCSC file` contains more cytoband calls._
+* `02-add-ploidy-consensus.Rmd` - This is very similar to the CNVkit file prep.
```
Comment (cansavvy) [inline analyses/focal-cn-file-preparation/README.md]: I am unsure what this is referring to:`similar to CNVKit file prep` can you put a link to what this is referring to?

File: analyses/focal-cn-file-preparation/README.md [hunk 6]
```diff
@@ -17,26 +15,31 @@ bash analyses/focal-cn-file-preparation/run-prepare-cn.sh
...
+However, there are instances in the consensus SEG file where `copy.num` is `NA` which are removed.
+See the notebook for more information.
+* `03-prepare-cn-file.R` - This script performs the ranges to annotation mapping using the GENCODE v27 GTF included via the data download step; it takes the ControlFreeC file or a SEG (e.g., CNVkit, consensus SEG) file prepared with `00-add-ploidy-cnvkit.Rmd` as input.
```
Comment (cansavvy) [inline analyses/focal-cn-file-preparation/README.md]: Unsure that what I put in this suggestion is right, but I see 00-add-ploidy-cnvkit.Rmd doesn't exist anymore. 
```suggestion
* `03-prepare-cn-file.R` - This script performs the ranges to annotation mapping using the GENCODE v27 GTF included via the data download step; it takes the ControlFreeC file or a SEG (e.g., CNVkit, consensus SEG) file prepared with `01-add-ploidy-cnvkit.Rmd` and  `02-add-ploidy-cnvkit.Rmd`as input.
```

File: analyses/focal-cn-file-preparation/README.md [hunk 7]
```diff
@@ -17,26 +15,31 @@ bash analyses/focal-cn-file-preparation/run-prepare-cn.sh
...
   A table with the following columns is returned:
   
-  | biospecimen_id | arm | broad_status | status | copy_number | ploidy | ensembl | gene_symbol |
-  |----------------|-----|--------------|--------|-------------|--------|---------|-------------|
+  | biospecimen_id | status | copy_number | ploidy | ensembl | gene_symbol | cytoband |
+  |----------------|--------|-------------|--------|---------|-------------|---------|
   
-* `02-rna-expression-validation.R` - This script examines RNA-seq expression levels (RSEM FPKM) of genes that are called as deletions.
+* `rna-expression-validation.R` - This script examines RNA-seq expression levels (RSEM FPKM) of genes that are called as deletions. 
+It is not currently run via the shell script.
```
Comment (cansavvy) [inline analyses/focal-cn-file-preparation/README.md]: What's the plan for this script then? Is it being dropped or revamped or is it supposed to be called somewhere else?

Comment (jaclyn-taroni) [reply analyses/focal-cn-file-preparation/README.md]: Plan is to rework it #387

File: analyses/focal-cn-file-preparation/02-add-ploidy-consensus.Rmd (hunk 2 above)
Comment (jaclyn-taroni) [reply analyses/focal-cn-file-preparation/02-add-ploidy-consensus.Rmd]: Ah, there was a moment where we thought the sex chromosomes were not represented in the consensus file at all. I was wrong and left in this comment.

File: analyses/focal-cn-file-preparation/README.md (hunk 5 above)
Comment (jaclyn-taroni) [reply analyses/focal-cn-file-preparation/README.md]: I added the `01` notebook file name here to make it more clear.

**Pull request merged**